### ISSUE TITLE
[apex] UnusedLocalVariable flags for variables which are using in SOQL/SOSL binds

### ISSUE THREAD

**Issue by nwcm** [pmd/pmd#4556]
**Affects PMD Version:** 7.0.0-rc2

**Rule:** [UnusedLocalVariable](https://docs.pmd-code.org/latest/pmd_rules_apex_bestpractices.html#unusedlocalvariable)

**Description:**

**Code Sample demonstrating the issue:**

```apex
    String unusedFlag = 'this string';
    String query = 'SELECT Id FROM Account WHERE Name =: unusedFlag';
```

**Expected outcome:**

PMD reports a violation at line 1, but that's wrong. That's a false positive.

**Running PMD through:** *[CLI]*

![image](https://github.com/pmd/pmd/assets/111259588/2ca312de-2106-479e-a926-114e962d1c5b)

**Comment by adangel** [comment]
Thanks for reporting.

The rule should support binding variables already. The implementation of the rule expects, that between the colon (`:`) and the variable name is _no_ space, you have a space, though. Is this still valid syntax?

```java
    String unusedFlag = 'this string';
    String query = 'SELECT Id FROM Account WHERE Name =: unusedFlag';
    //                                                  ^
    //                                                  | is this space allowed?
```

**Comment by adangel** [comment]
Oh, and one more thing: You don't use the string yet in a query, so PMD can't know, that this string is supposed to be a SOQL/SOSL query...
You need at least something like `Database.query(query);`

See also #2669

**Comment by nwcm** [comment]
@adangel Thanks for the comments, 

Sorry I must have missed it in my example. Regardless of `=:` or `=: ` which are both valid at least in v56 as below it will report unused

![image](https://github.com/pmd/pmd/assets/111259588/b4eee4f8-c069-4098-ae62-8649f8d567da)

**Comment by nwcm** [comment]
If we update the BINDING_VARIABLE regex to `(?i):(?:\s|)([_a-z0-9]+)` it will capture all the possible 

I have yet to test it, but i have opened a draft PR for it. Though even `=:foo` does incorrect throw unused variable in 7.0.0rc2

https://github.com/pmd/pmd/pull/4562

**Comment by adangel** [comment]
Note: Your screenshot shows, that it's never working - with or without a space. The problem seems to be, that https://github.com/ChuckJonas/vscode-apex-pmd still uses PMD 6.42.0 (see https://github.com/ChuckJonas/vscode-apex-pmd/tree/master/bin/pmd/lib) and support for binding vars in dynamic queries was added only with PMD 6.54.0 (see #2669).

> Though even =:foo does incorrect throw unused variable in 7.0.0rc2

That I can't reproduce. Are you sure, you are using PMD 7.0.0-rc2?

The false positive with using a space before the colon is still valid though and needs to be fixed...

**Comment by nwcm** [comment]
Good catch @adangel I was on another device and I hadn't configured the pmd bin path. 

So this issue and suggestions are more for variable names with `_` and the space in bindings

![image](https://github.com/pmd/pmd/assets/111259588/152359b0-0cb8-4c82-b5a9-58e6ae8a3475)
